lib/src/mt_ptp.c

```diff
@@ -377,7 +377,32 @@ static void ptp_calculate_coefficient(struct mt_ptp_impl* ptp, int64_t delta) {
 }
 
 static void ptp_adjust_delta(struct mt_ptp_impl* ptp, int64_t delta) {
-  ptp_timesync_adjust_time(ptp, delta);
+  double ppb;
+  enum servo_state state = UNLOCKED;
+
+  if (mt_has_phc2sys_service(ptp->impl)) {
+    ppb = pi_sample(&ptp->servo, -1 * delta, ptp->t2, &state);
+
+    switch (state) {
+      case UNLOCKED:
+        break;
+      case JUMP:
+        ptp_timesync_adjust_time(ptp, delta);
+        dbg("%s(%d), master offset: %" PRId64 " path delay: %" PRId64 " adjust time.\n",
+            __func__, ptp->port_id, delta, ptp->path_delay);
+        break;
+      case LOCKED:
+        ptp_timesync_lock(ptp);
+        rte_eth_timesync_adjust_freq(ptp->port_id, -1 * (long)(ppb * 65.536));
```

Comment:
Can you wrap it to something like `ptp_timesync_adjust_freq`, and return fail if it's no_timesync case. And `rte_eth_timesync_adjust_freq` is one new API not exist for all DPDK versions, we may need find a way to still pass the build if the API is not there.

Reply:
> Can you wrap it to something like `ptp_timesync_adjust_freq`, and return fail if it's no_timesync case. And `rte_eth_timesync_adjust_freq` is one new API not exist for all DPDK versions, we may need find a way to still pass the build if the API is not there.

Used macro to enable freq adjust only based on dpdk 23.03 with the private patch `0007-igc-optimize-LaunchTime-Tx-Qbv-configuration-and-PTP.patch`.

Reply:
No, we can't use DPDK version marco since the patch is not going into the DPDP upstream code really. Can you refer to https://github.com/OpenVisualCloud/Media-Transport-Library/blob/main/patches/dpdk/23.03/hdr_split/0001-net-ice-support-hdr-split-mbuf-callback.patch#L188C10-L188C19 to add a marco like 'MTL_HAS_DPDK_TIMESYNC_ADJUST_FREQ' into `0007-igc-optimize-LaunchTime-Tx-Qbv-configuration-and-PTP.patch`

Reply:
> No, we can't use DPDK version marco since the patch is not going into the DPDK upstream code really. Can you refer to https://github.com/OpenVisualCloud/Media-Transport-Library/blob/main/patches/dpdk/23.03/hdr_split/0001-net-ice-support-hdr-split-mbuf-callback.patch#L188C10-L188C19 to add a marco like 'MTL_HAS_DPDK_TIMESYNC_ADJUST_FREQ' into `0007-igc-optimize-LaunchTime-Tx-Qbv-configuration-and-PTP.patch`

Done.

Reply:
Can you help to warp `rte_eth_timesync_adjust_freq` to `ptp_timesync_adjust_freq`, and add error handling if the adjust action fail. E810 NIC has the `MTL_HAS_DPDK_TIMESYNC_ADJUST_FREQ` patch but device is not capable. And is there a way to detect if NIC is capable of adjust freq at runtime?

Reply:
> Can you help to warp `rte_eth_timesync_adjust_freq` to `ptp_timesync_adjust_freq`, and add error handling if the adjust action fail. E810 NIC has the `MTL_HAS_DPDK_TIMESYNC_ADJUST_FREQ` patch but device is not capable. And is there a way to detect if NIC is capable of adjust freq at runtime?

Done.